### PR TITLE
support min_p sampling & do_sample setting

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

Support `min_p` sampling and add do_sample setting.

related issue https://github.com/InternLM/lmdeploy/issues/1745

| batch | vocab | topk | before | after + without minp | after + with minp |
| ----- | ----- | ---- | ------ | -------------------- | ----------------- |
| 32    | 96000 | 40   | 0.1782 | 0.1759               | 0.1767            |
| 32    | 96000 | 1024 |        | 4.11                 | 4.11              |
|       |       |      |        |                      |                   |
|       |       |      |        |                      |                   |
|       |       | topp | before | after + without minp | after + with minp |
| 32    | 96000 | 1    | 1.0456 | 1.0442               | 1.0415            |

---

The current implementation of the sampling layer has some shortcomings. 
1. The filtering and sampling processes are writern in one kernel that makes it inconvenient to add new filters. 
2. There are some operation doesn't need to be called twice like `RepetitionPenalty`, `TemperaturePenalty` etc. 
3. The implementation of topk kernnel needs to be modified when the value of topk is large. The time should't be much higher then topp.

~~I will submit another PR to fix these issues, this PR will focus on the min_p functionality.~~

### CODE REVIEW

**Comment by zhyncs** [comment]
Looking forward to the `do_sample` parameter very much 😂

**Comment by zhyncs** [comment]
After the merge of https://github.com/InternLM/lmdeploy/pull/1912, will it be better to update this pull request on that basis and then review it?

**Comment by irexyc** [comment]
> After the merge of https://github.com/InternLM/lmdeploy/pull/1912, will it be better to update this pull request on that basis and then review it?

ok

**Comment by lzhangzz** [comment]
@irexyc is this still WIP？

File: src/turbomind/kernels/sampling_topk_kernels.cu [hunk 1]
```diff
@@ -283,13 +298,60 @@ __global__ void topk_stage2_sampling(const int* __restrict topk_tmp_id_buf,
         __syncthreads();
     }
 
-    int selected_i;
+    // top_p
```
Comment (lvhan028) [inline src/turbomind/kernels/sampling_topk_kernels.cu:301]: top_p?

**Pull request closed without merging**